src/dbnode/storage/bootstrap/bootstrapper/base.go

```diff
@@ -76,133 +74,261 @@ func (b baseBootstrapper) String() string {
+		if err != nil {
+			return bootstrap.NamespaceResults{}, err
+		}
+
+		currNamespace.DataRunOptions.ShardTimeRanges = dataAvailable
+
+		// Prepare index if required.
+		if namespace.Metadata.Options().IndexOptions().Enabled() {
+			indexAvailable, err := b.src.AvailableIndex(namespace.Metadata,
+				namespace.DataRunOptions.ShardTimeRanges.Copy(),
+				namespace.DataRunOptions.RunOptions)
+			if err != nil {
+				return bootstrap.NamespaceResults{}, err
+			}
+
+			currNamespace.IndexRunOptions.ShardTimeRanges = indexAvailable
+		}
+
+		// Set the namespace options for the current bootstrapper source.
+		curr.Namespaces.Set(id, currNamespace)
+
+		// Log the metadata about bootstrapping this namespace.
+		dataRange = currNamespace.DataRunOptions.ShardTimeRanges.MinMaxRange()
+		prepareLogFields := append(logFieldsCopy(logFields), []zapcore.Field{
+			zap.String("namespace", id.String()),
+			zap.Int("numShards", len(currNamespace.DataRunOptions.ShardTimeRanges)),
+			zap.Duration("dataRange", dataRange),
+		}...)
+		if dataRange > 0 {
+			dataMin, dataMax := currNamespace.DataRunOptions.ShardTimeRanges.MinMax()
+			prepareLogFields = append(prepareLogFields, []zapcore.Field{
+				zap.Time("dataFrom", dataMin),
+				zap.Time("dataTo", dataMax),
+			}...)
+		}
+
+		if namespace.Metadata.Options().IndexOptions().Enabled() {
+			indexRange := currNamespace.IndexRunOptions.ShardTimeRanges.MinMaxRange()
+			prepareLogFields = append(prepareLogFields, []zapcore.Field{
+				zap.Duration("indexRange", indexRange),
+			}...)
+			if indexRange > 0 {
+				indexMin, indexMax := currNamespace.IndexRunOptions.ShardTimeRanges.MinMax()
+				prepareLogFields = append(prepareLogFields, []zapcore.Field{
+					zap.Time("indexFrom", indexMin),
+					zap.Time("indexTo", indexMax),
+				}...)
+			}
+		}
+
+		b.log.Info("bootstrap from source for namespace ready", prepareLogFields...)
 	}
-	b.log.Info("bootstrapping from source starting", logFields...)
 
 	nowFn := b.opts.ClockOptions().NowFn()
 	begin := nowFn()
 
-	currStatus, currErr = step.runCurrStep(currRanges)
+	b.log.Info("bootstrap from source started", logFields...)
+	currResults, err := b.src.Read(curr)
 
 	logFields = append(logFields, zap.Duration("took", nowFn().Sub(begin)))
-	if currErr != nil {
-		logFields = append(logFields, zap.Error(currErr))
-		b.log.Info("bootstrapping from source completed with error", logFields...)
-	} else {
-		logFields = append(logFields, currStatus.logFields...)
-		b.log.Info("bootstrapping from source completed successfully", logFields...)
+	if err != nil {
+		errorLogFields := append(logFieldsCopy(logFields), zap.Error(err))
+		b.log.Info("bootstrapping from source completed with error", errorLogFields...)
+	}
+	if err != nil {
+		return bootstrap.NamespaceResults{}, err
 	}
 
-	wg.Wait()
-	if err := xerrors.FirstError(currErr, nextErr); err != nil {
-		return err
+	// Determine the unfulfilled and the unattempted ranges to execute next.
```

Comment:
nit: any chance can we split this out into a separate function for readability?

Reply:
Sounds good, done.